### PR TITLE
Flatten all interface variables

### LINKED ISSUE #1660: input/output mismatch of HLSL vertex shader and geometry shader
The following vertex and geometry shader works well on D3D, but generated some validation errors for Vulkan.

```
struct VS_STAR_INPUT
{
	float4 iPos : POSITION0;
	float4 iColor : COLOR0;
};

struct VS_STAR_OUTPUT
{
	float4 oPos : POSITION0;
	float4 oColor : COLOR0;
	float2 oSize : TEXCOORD1;
};

VS_STAR_OUTPUT vs_main_star(VS_STAR_INPUT input)
{
	VS_STAR_OUTPUT output;
	output.oPos = input.iPos;
	output.oColor = float4(1.0, 1.0, 1.0, 1.0);
	output.oSize = float2(1.0, 1.0);
	
	return output;
}

struct GS_STAR_OUTPUT
{
	float4 oPos : SV_POSITION;
	float2 oTex	: TEXCOORD0;
	float4 oColor : COLOR0;
};

[maxvertexcount (4)]
void gs_main_star(point VS_STAR_OUTPUT In[1], inout TriangleStream<GS_STAR_OUTPUT> triStream)
{
	for(int v = 0; v < 4; v ++)
	{   
		GS_STAR_OUTPUT Out;
	
		Out.oPos = In[0].oPos * In[0].oSize.x;
		Out.oTex = float2(1.0, 1.0);
		Out.oColor = In[0].oColor;
	
		triStream.Append(Out);
	}
	triStream.RestartStrip();
}
```

```
Type mismatch on location 0.0: 'ptr to output vec4 of float32' vs 'ptr to input arr[1] of struct of (vec4 of float32, vec4 of float32, vec2 of float32)'
vertex shader writes to output location 1.0 which is not consumed by geometry shader
vertex shader writes to output location 2.0 which is not consumed by geometry shader
```

Looks like the generated spir-v of geometry shader doesn't flatten the input struct, so the vertex shader has 3 output, but geometry shader only have 1 input.

### PR DESCRIPTION
**Pull request by greg-lunarg**
Specifically, add flattening of arrayed io for geometry and
tesselation shaders. Previously some interface structs just had
builtins split out which caused some interfaces to not be exactly
the same as that of flattened adjacent stages, affecting validation
and correctness.

This obviates builtin splitting. That will be removed in a followup
commit. It was left in for this commit to better exhibit the functional
changes that were made.

Fixes #1660.

### CODE REVIEW

**Comment by greg-lunarg** [comment]
FWIW, I visually inspected all changed SPIR-V.

**Comment by greg-lunarg** [comment]
@danginsburg also verified correctness with his internal testing which tests geometry and tesselation as well as verified that the problem he reported in #1660 was now passing validation and rendering correctly.

**Comment by danginsburg** [comment]
Yes, this change was run through our hundreds of thousands of shaders.  It caused no differences on VS/PS, and only changed GS/HS/DS.  It confirmed fix the PSO creation validation error and also fixed a rendering bug with it on RADV.  Thanks @greg-lunarg!

**Comment by danginsburg** [comment]
I wonder if this also fixes 

https://github.com/KhronosGroup/glslang/issues/1738
https://github.com/KhronosGroup/glslang/issues/1737

File: hlsl/hlslParseHelper.cpp [hunk 1]
```diff
@@ -1181,8 +1181,14 @@ void HlslParseContext::flatten(const TVariable& variable, bool linkage)
                                                   TFlattenData(type.getQualifier().layoutBinding,
                                                                type.getQualifier().layoutLocation)));
 
-    // the item is a map pair, so first->second is the TFlattenData itself.
-    flatten(variable, type, entry.first->second, variable.getName(), linkage, type.getQualifier(), nullptr);
+    // if flattening arrayed io struct, array each member of dereferenced type
+    if (arrayed) {
+        const TType dereferencedType(type, 0);
+        flatten(variable, dereferencedType, entry.first->second, variable.getName(), linkage, type.getQualifier(), type.getArraySizes());
+    }
```
Comment (johnkslang) [inline hlsl/hlslParseHelper.cpp:1188]: Follow existing conventions.  `} else {`

File: hlsl/hlslParseHelper.cpp [hunk 2]
```diff
@@ -1256,6 +1262,10 @@ int HlslParseContext::addFlattenedMember(const TVariable& variable, const TType&
             }
         }
 
+        // Only propagate arraysizes here for arrayed io
+        if (variable.getType().getQualifier().isArrayedIo(language) && builtInArraySizes != nullptr)
+          memberVariable->getWritableType().copyArraySizes(*builtInArraySizes);
```
Comment (johnkslang) [inline hlsl/hlslParseHelper.cpp:1267]: Glslang is still using 4-space indentation.

File: hlsl/hlslParseHelper.cpp [hunk 3]
```diff
@@ -2733,17 +2740,33 @@ TIntermTyped* HlslParseContext::handleAssign(const TSourceLoc& loc, TOperator op
                wasSplit(binaryNode->getLeft());
     };
 
+    // Return symbol if node is symbol or index ref
+    const auto getSymbol = [](const TIntermTyped* node) -> const TIntermSymbol* {
+      const TIntermSymbol* symbolNode = node->getAsSymbolNode();
```
Comment (johnkslang) [inline hlsl/hlslParseHelper.cpp:2745]: Need 4-space indentation throughout.

File: hlsl/hlslParseHelper.cpp [hunk 4]
```diff
@@ -2869,10 +2894,36 @@ TIntermTyped* HlslParseContext::handleAssign(const TSourceLoc& loc, TOperator op
...
             }
         } else if (flattened && !shouldFlatten(derefType, isLeft ? leftStorage : rightStorage, false)) {
-            if (isLeft)
+            if (isLeft) {
+                // offset will cycle through variables for arrayed io
+                if (leftOffset >= static_cast<int>(leftVariables->size()))
+                    leftOffset = leftOffsetStart;
                 subTree = intermediate.addSymbol(*(*leftVariables)[leftOffset++]);
-            else
+            }
```
Comment (johnkslang) [inline hlsl/hlslParseHelper.cpp:2902]: `} else {` throughout.

**Comment by johnkslang** [comment]
Yes, eliminating split sounds very familiar, I believe most the work to do that was already done, and this is the last part.

Basically, this looks okay, just need to address the nits about coding conventions.

**Comment by greg-lunarg** [comment]
> I wonder if this also fixes

> #1738
> #1737

Once this is merged I will close out all the related issues.

**Comment by greg-lunarg** [comment]
I believe I have fixed all style issues. Sorry about that!

**Pull request merged**